sources/api/migration/migrator/src/main.rs

```diff
@@ -580,60 +476,193 @@ where
+        std::os::unix::fs::symlink(&datastore_major, &datastore_current).unwrap();
+    }
+
+    fn root() -> PathBuf {
+        test_data()
+            .join("repository")
+            .join("metadata")
+            .join("1.root.json")
+    }
 
-        // Test no matching migrations
-        assert!(select_migrations(&v04, &v05, &all_migrations)
-            .unwrap()
-            .is_empty());
+    fn tuf_metadata() -> PathBuf {
+        test_data().join("repository").join("metadata")
+    }
+
+    fn tuf_targets() -> PathBuf {
+        test_data().join("repository").join("targets")
+    }
+
+    /// Tests the migrator program end-to-end using the `run` function.
+    /// The test uses a locally stored tuf repo at `migrator/tests/data/repository`.
+    /// In the `manifest.json` we have specified the following migrations:
+    /// ```
+    ///     "(0.99.0, 0.99.1)": [
+    ///       "x-first-migration.lz4",
+    ///       "a-second-migration.lz4"
+    ///     ]
+    /// ```
+    ///
+    /// The two 'migrations' are bash scripts with content like this:
+    ///
+    /// ```
+    /// #!/bin/bash
+    /// set -eo pipefail
+    /// migration_name="x-first-migration"
+    /// datastore_parent_dir="$(dirname "${3}")"
+    /// outfile="${datastore_parent_dir}/result.txt"
+    /// echo "${migration_name}: writing a message to '${outfile}'"
+    /// echo "${migration_name}:" "${@}" >> "${outfile}"
+    /// ```
+    ///
+    /// These 'migrations' use the --source-datastore path and take its parent.
+    /// Into this parent directory they write lines to a file named result.txt.
+    /// In the test we read the result.txt file to see that the migrations have been run in the
+    /// expected order.
+    ///
+    /// This test ensures that migrations run when migrating from an older to a newer version.
+    #[test]
+    fn migrate_forward() {
+        let from_version = Version::parse("0.99.0").unwrap();
+        let to_version = Version::parse("0.99.1").unwrap();
+        let mut info = MigrationTestInfo::new(from_version, to_version);
+        create_datastore_links(&mut info);
+        let args = Args {
+            datastore_path: info.datastore.clone(),
+            log_level: log::LevelFilter::Info,
+            migration_directory: tuf_targets(),
+            migrate_to_version: info.to_version.clone(),
+            root_path: root(),
+            metadata_directory: tuf_metadata(),
+        };
+        run(&args).unwrap();
+        // the migrations should write to a file named result.txt.
+        let output_file = info.tmp.path().join("result.txt");
+        let contents = std::fs::read_to_string(&output_file).unwrap();
+        let lines: Vec<&str> = contents.split('\n').collect();
+        assert_eq!(lines.len(), 3);
+        let first_line = *lines.get(0).unwrap();
+        if !first_line.contains("x-first-migration: --forward") {
+            panic!(format!(
```

Comment:
Can we simplify these checks using `assert!` instead of `if` and `panic`?

Reply:
Addressed by:

> webern force-pushed the webern:pr/signed-migrations branch from f481a2d to 4443441